### PR TITLE
use own SA for each component

### PR DESCRIPTION
**Pull request by stanlyDoge**
### Type of change
- Enhancement / new feature
### Description
Each component (zk, mirror maker, connect, ..) should have its own SA.

After deploying example cluster + mirror maker&kafka connect

```[sknot@sknot mockkube]$ oc get sa
NAME                           SECRETS   AGE
builder                        2         3m
default                        2         3m
deployer                       2         3m
my-cluster-entity-operator     2         50s
my-cluster-kafka               2         1m
my-cluster-zookeeper           2         1m
my-connect-cluster-connect     2         13s
my-mirror-maker-mirror-maker   2         25s
strimzi-cluster-operator       2         2m
 ```

### Checklist

- [ ] Update/write design documentation in `./design`
- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md

### CODE REVIEW

**Comment by ppatierno** [comment]
@stanlyDoge I see 148 files changed, maybe a wrong rebase? :)

**Comment by stanlyDoge** [comment]
Yeah, it is just draft PR. It is based on https://github.com/strimzi/strimzi-kafka-operator/pull/1346 so I definitely need to rebase later :)

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/Main.java [hunk 1]
```diff
@@ -116,7 +117,7 @@ static CompositeFuture run(Vertx vertx, KubernetesClient client, boolean isOpenS
                 config.versions(), config.getImagePullPolicy());
         KafkaConnectAssemblyOperator kafkaConnectClusterOperations = new KafkaConnectAssemblyOperator(vertx, isOpenShift,
                 certManager, kco, configMapOperations, deploymentOperations, serviceOperations, secretOperations,
-                networkPolicyOperator, podDisruptionBudgetOperator, config.versions(), config.getImagePullPolicy());
+                networkPolicyOperator, podDisruptionBudgetOperator, new ResourceOperatorSupplier(vertx, client, isOpenShift, config.getOperationTimeoutMs()), config.versions(), config.getImagePullPolicy());
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/Main.java]: Isn't there an existing instance you can reuse?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/Main.java [hunk 2]
```diff
@@ -129,7 +130,7 @@ static CompositeFuture run(Vertx vertx, KubernetesClient client, boolean isOpenS
         KafkaMirrorMakerAssemblyOperator kafkaMirrorMakerAssemblyOperator =
                 new KafkaMirrorMakerAssemblyOperator(vertx, isOpenShift, certManager, kmmo, secretOperations,
                         configMapOperations, networkPolicyOperator, deploymentOperations, serviceOperations,
-                        podDisruptionBudgetOperator, config.versions(), config.getImagePullPolicy());
+                        podDisruptionBudgetOperator, new ResourceOperatorSupplier(vertx, client, isOpenShift, config.getOperationTimeoutMs()), config.versions(), config.getImagePullPolicy());
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/Main.java]: Isn't there an existing instance you can reuse?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/Main.java [hunk 3]
```diff
@@ -233,15 +234,24 @@ private static KafkaConnectS2IAssemblyOperator createS2iOperator(Vertx vertx, Ku
                     put("strimzi-kafka-broker", "030-ClusterRole-strimzi-kafka-broker.yaml");
                     put("strimzi-entity-operator", "031-ClusterRole-strimzi-entity-operator.yaml");
                     put("strimzi-topic-operator", "032-ClusterRole-strimzi-topic-operator.yaml");
+                    put("strimzi-zookeeper-broker", "033-ClusterRole-strimzi-zookeeper-broker.yaml");
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/Main.java]: zookeeper isn't a broker, so filename is misleading.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/Main.java [hunk 4]
```diff
@@ -233,15 +234,24 @@ private static KafkaConnectS2IAssemblyOperator createS2iOperator(Vertx vertx, Ku
...
+                    put("strimzi-kafka-mirror-maker-broker", "034-ClusterRole-strimzi-mirror-maker-broker.yaml");
+                    put("strimzi-kafka-connect-broker", "035-ClusterRole-strimzi-connect-broker.yaml");
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/Main.java]: Same here, kafka connect instances aren't brokers.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaMirrorMakerCluster.java [hunk 5]
```diff
@@ -451,4 +459,59 @@ private void setProducerUsernameAndPassword(String username, PasswordSecretSourc
...
+                .withKind("ClusterRole")
+                .build();
+
+        return new KubernetesClusterRoleBindingBuilder()
+                .withNewMetadata()
+                .withName(initContainerClusterRoleBindingName(namespace, cluster))
+                .withNamespace(assemblyNamespace)
+                .withOwnerReferences(createOwnerReference())
+                .withLabels(labels.toMap())
+                .endMetadata()
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaMirrorMakerCluster.java]: As a general comment please indent the code between `withNew*()` and `end*()`, that way the indentation reflects the nesting structure of the objects being built and makes the code much easier to read.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java [hunk 6]
```diff
@@ -880,6 +883,34 @@ ReconciliationState createReconciliationState(Reconciliation reconciliation, Kaf
...
+            KubernetesClusterRoleBinding desired = zkCluster.generateClusterRoleBinding(namespace);
+            Future<ReconcileResult<KubernetesClusterRoleBinding>> fut = clusterRoleBindingOperator.reconcile(
+                    ZookeeperCluster.initContainerClusterRoleBindingName(namespace, name), desired);
+
+            Future replacementFut = Future.future();
+
+            fut.setHandler(res -> {
+                if (res.failed()) {
+                    if (desired == null) {
+                        replacementFut.complete();
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: Why is this necessary?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperator.java [hunk 7]
```diff
@@ -118,4 +130,34 @@ public KafkaConnectAssemblyOperator(Vertx vertx, boolean isOpenShift,
...
+        KubernetesClusterRoleBinding desired = connect.generateClusterRoleBinding(namespace);
+        Future<ReconcileResult<KubernetesClusterRoleBinding>> fut = clusterRoleBindingOperations.reconcile(
+                KafkaConnectCluster.initContainerClusterRoleBindingName(namespace, name), desired);
+
+        Future replacementFut = Future.future();
+
+        fut.setHandler(res -> {
+            if (res.failed()) {
+                if (desired == null) {
+                    replacementFut.complete();
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperator.java]: Again, I don't understand why a failure when `desired == null` is OK to ignore?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMakerAssemblyOperator.java [hunk 8]
```diff
@@ -61,7 +68,7 @@
      * @param configMapOperations        For operating on ConfigMaps
      * @param secretOperations           For operating on Secrets
      */
-    public KafkaMirrorMakerAssemblyOperator(Vertx vertx, boolean isOpenShift,
+    public KafkaMirrorMakerAssemblyOperator(Vertx vertx,  boolean isOpenShift,
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMakerAssemblyOperator.java]:
```suggestion
    public KafkaMirrorMakerAssemblyOperator(Vertx vertx, boolean isOpenShift,
```

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMakerAssemblyOperator.java [hunk 9]
```diff
@@ -123,4 +135,33 @@ public KafkaMirrorMakerAssemblyOperator(Vertx vertx, boolean isOpenShift,
...
+                    replacementFut.complete();
+                } else {
+                    replacementFut.fail(res.cause());
+                }
+            } else {
+                replacementFut.complete();
+            }
+        });
+
+        return replacementFut;
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMakerAssemblyOperator.java]: This whole function seems to be repeated three times in this PR. Is there some way you can factor it out into a common method/function somewhere?

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperatorTest.java [hunk 10]
```diff
@@ -81,6 +95,51 @@ public static void after() {
...
+                    .withIngress(new LoadBalancerIngressBuilder().withHostname(i.getArgument(0) + "." + i.getArgument(1) + ".mydomain.com").build())
+                    .endLoadBalancer()
+                    .endStatus()
+                    .withNewSpec()
+                    .withPorts(new ServicePortBuilder().withNodePort(31245).build())
+                    .endSpec()
+                    .build();
+        });
+
+        return supplier;
```
Comment (tombentley) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperatorTest.java]: This seems to be a copy of the code in `KafkaConnectAssemblyOperatorMockTest`. Can you factor it into a common method somewhere? ResourceUtils might be a good place.

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMakerAssemblyOperatorTest.java [hunk 11]
```diff
@@ -87,6 +101,51 @@ public static void after() {
...
+                    .withIngress(new LoadBalancerIngressBuilder().withHostname(i.getArgument(0) + "." + i.getArgument(1) + ".mydomain.com").build())
+                    .endLoadBalancer()
+                    .endStatus()
+                    .withNewSpec()
+                    .withPorts(new ServicePortBuilder().withNodePort(31245).build())
+                    .endSpec()
+                    .build();
+        });
+
+        return supplier;
```
Comment (tombentley) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMakerAssemblyOperatorTest.java]: And here is the same code again!

File: operator-common/src/main/java/io/strimzi/operator/common/operator/resource/ClusterRoleOperator.java [hunk 12]
```diff
@@ -4,59 +4,87 @@
...
-        return urlWithoutName() + "/" + name;
+    @Override
+    protected Future<ReconcileResult<KubernetesClusterRole>> internalPatch(String name, KubernetesClusterRole current, KubernetesClusterRole desired) {
+        return Future.succeededFuture(ReconcileResult.noop(current));
     }
 
-    public Future<Void> reconcile(String name, ClusterRole resource) {
-        return doReconcile(urlWithoutName(), urlWithName(name), resource);
-    }
+    public static KubernetesClusterRole convertYamlToClusterRole(String yaml) {
```
Comment (tombentley) [inline operator-common/src/main/java/io/strimzi/operator/common/operator/resource/ClusterRoleOperator.java]: I don't understand why you needed to write this method. It doesn't seem to be the best way of doing whatever is needed? In particular why can't you just use something like `yamlReader.readValue(yaml, KubernetesClusterRole.class)`?

File: operator-common/src/main/java/io/strimzi/operator/common/operator/resource/AbstractNonNamespacedResourceOperator.java [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+    /**
+     * Returns a future that completes when the resource identified by the given {@code name}
+     * is ready.
+     *
+     * @param name The resource name.
+     * @param pollIntervalMs The poll interval in milliseconds.
+     * @param timeoutMs The timeout, in milliseconds.
+     * @param predicate The predicate.
+     */
+    public Future<Void> waitFor(String name, long pollIntervalMs, final long timeoutMs, Predicate<String> predicate) {
```
Comment (tombentley) [inline operator-common/src/main/java/io/strimzi/operator/common/operator/resource/AbstractNonNamespacedResourceOperator.java]: Why does this need to be a `public` method in this class? It seems to add very little on top of `Util.waitFor()`. Does it need to be overridden in a subclass? How many callers does it have? Can't you just inline it?

File: .checkstyle/suppressions.xml [hunk 14]
```diff
@@ -24,6 +24,9 @@
     <suppress checks="ClassFanOutComplexity|JavaNCSS"
               files="io[/\\]strimzi[/\\]operator[/\\]cluster[/\\]operator[/\\]assembly[/\\]KafkaAssemblyOperatorTest.java"/>
 
+    <suppress checks="ClassFanOutComplexity"
```
Comment (scholzj) [inline .checkstyle/suppressions.xml]: Should we use the annotation inside the class instead?

File: install/cluster-operator/033-ClusterRole-strimzi-zookeeper-cluster.yaml [hunk 15]
```diff
@@ -0,0 +1,13 @@
+apiVersion: rbac.authorization.k8s.io/v1beta1
+kind: ClusterRole
+metadata:
+  name: strimzi-zookeeper-cluster
```
Comment (scholzj) [inline install/cluster-operator/033-ClusterRole-strimzi-zookeeper-cluster.yaml]: We do not need this role and Zookeeper doesn't need these rights. The same applies to mirror maker and connect. You should remove these new RBAC files and just create the Service Accounts. This is just about isolating the applications by giving them service account without any actual rights.

Comment (stanlyDoge) [reply install/cluster-operator/033-ClusterRole-strimzi-zookeeper-cluster.yaml]: Ah, thanks. Should be fixed now.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/AbstractModel.java [hunk 16]
```diff
@@ -1078,4 +1080,19 @@ public static String clusterCaKeySecretName(String cluster)  {
 
         return merged;
     }
+
+    public ServiceAccount generateServiceAccountInternal() {
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/AbstractModel.java]: Why do we have these two methods? Why is this not part of the `generateServiceAccount` directly?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaConnectCluster.java [hunk 17]
```diff
@@ -535,4 +535,16 @@ protected void setUsernameAndPassword(String username, PasswordSecretSource pass
...
+
+    @Override
+    protected String getServiceAccountName() {
+        return initContainerServiceAccountName(cluster);
+    }
+
+    /**
+     * Get the name of the connect service account given the name of the {@code connectResourceName}.
+     */
+    public static String initContainerServiceAccountName(String connectResourceName) {
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaConnectCluster.java]: This probably should be renamed. It is called `initContainer...` because in Kafka it was used mainly because of the init container. But there are no init containers here, Zookeeper or in Mirror Maker.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaMirrorMakerCluster.java [hunk 18]
```diff
@@ -454,4 +454,16 @@ private void setProducerUsernameAndPassword(String username, PasswordSecretSourc
...
+
+    @Override
+    protected String getServiceAccountName() {
+        return initContainerServiceAccountName(cluster);
+    }
+
+    /**
+     * Get the name of the mirror maker service account given the name of the {@code mirrorMakerResourceName}.
+     */
+    public static String initContainerServiceAccountName(String mirrorMakerResourceName) {
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaMirrorMakerCluster.java]: Same as above.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/ZookeeperCluster.java [hunk 19]
```diff
@@ -479,4 +479,16 @@ protected void setTlsSidecar(TlsSidecar tlsSidecar) {
     protected String getDefaultLogConfigFileName() {
         return "zookeeperDefaultLoggingProperties";
     }
+
+    /**
+     * Get the name of the zookeeper service account given the name of the {@code zookeeperResourceName}.
+     */
+    public static String initContainerServiceAccountName(String zookeeperResourceName) {
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/ZookeeperCluster.java]: Same as with Connect and MM.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java [hunk 20]
```diff
@@ -167,6 +167,7 @@ public KafkaAssemblyOperator(Vertx vertx, boolean isOpenShift,
                 .compose(state -> state.zkManualPodCleaning())
                 .compose(state -> state.zkManualRollingUpdate())
                 .compose(state -> state.getZookeeperDescription())
+                .compose(state -> state.zookeeperInitServiceAccount())
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: Again, it should not be called `Init`

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/AbstractModel.java (hunk 16 above)
Comment (stanlyDoge) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/model/AbstractModel.java]: In `EntityOperator` we perform `isDeployed` check so we need to use this like that.

Comment (scholzj) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/model/AbstractModel.java]: Ok, lets see how @tombentley likes it ... maybe he proposes a better solution :-)

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/Main.java [hunk 21]
```diff
@@ -240,7 +240,12 @@ private static KafkaConnectS2IAssemblyOperator createS2iOperator(Vertx vertx, Ku
             for (Map.Entry<String, String> clusterRole : clusterRoles.entrySet()) {
                 log.info("Creating cluster role {}", clusterRole.getKey());
 
-                try (BufferedReader br = new BufferedReader(new InputStreamReader(Main.class.getResourceAsStream("/cluster-roles/" + clusterRole.getValue()), Charset.defaultCharset()))) {
+                try (BufferedReader br = new BufferedReader(
+                        new InputStreamReader(
+                                Main.class.getResourceAsStream(
+                                        "/cluster-roles/"
+                                                + clusterRole.getValue()),
+                                Charset.defaultCharset()))) {
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/Main.java]: Its very unlikely to be a problem, but this ends up being completely undefined behavior. Think about it: There's this file which gets edited on a developers machine (using who knows what char encoding), packaged into a jar, embedded in an image and then parsed at runtime using the default charset of the container. What ensures that the charset the developer's text editor or IDE used was the same as the default charset of the container? Do you even know what the default chartset of the container is, or that it doesn't very between nodes? I think it's much better to use UTF-8 here, and for all the YAMLs, that way at least you know for sure what you're reading the file as. (The other part of the problem is knowing what charset developers are using, but that's not really something you can solve in this PR).

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/AbstractModel.java (hunk 16 above)
Comment (tombentley) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/model/AbstractModel.java]: I think in EO you can just override

    @Override
    public ServiceAccount generateServiceAccount() {

        if (!isDeployed()) {
            return null;
        }
        return super.generateServiceAccount()
    }

So there's no need for the internal method.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperator.java [hunk 22]
```diff
@@ -118,4 +127,11 @@ public KafkaConnectAssemblyOperator(Vertx vertx, boolean isOpenShift,
     protected List<HasMetadata> getResources(String namespace, Labels selector) {
         return Collections.EMPTY_LIST;
     }
+
+
+    Future connectServiceAccount(String namespace, KafkaConnectCluster connect) {
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperator.java]: `Future` is generic so please give type arguments.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectS2IAssemblyOperator.java [hunk 23]
```diff
@@ -136,4 +146,9 @@ public KafkaConnectS2IAssemblyOperator(Vertx vertx, boolean isOpenShift,
         return Collections.EMPTY_LIST;
     }
 
+    Future connectServiceAccount(String namespace, KafkaConnectCluster connect) {
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectS2IAssemblyOperator.java]: Future is generic

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMakerAssemblyOperator.java [hunk 24]
```diff
@@ -123,4 +132,10 @@ public KafkaMirrorMakerAssemblyOperator(Vertx vertx, boolean isOpenShift,
         
         return result;
     }
+
+    Future mirrorMakerServiceAccount(String namespace, KafkaMirrorMakerCluster mirror) {
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMakerAssemblyOperator.java]: Future is generic

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperatorMockTest.java [hunk 25]
```diff
@@ -278,6 +278,7 @@ private KafkaAssemblyOperator createCluster(TestContext context) {
             context.assertNotNull(mockClient.secrets().inNamespace(NAMESPACE).withName(KafkaCluster.brokersSecretName(CLUSTER_NAME)).get());
             context.assertNotNull(mockClient.secrets().inNamespace(NAMESPACE).withName(ZookeeperCluster.nodesSecretName(CLUSTER_NAME)).get());
             context.assertNotNull(mockClient.secrets().inNamespace(NAMESPACE).withName(TopicOperator.secretName(CLUSTER_NAME)).get());
+            context.assertNotNull(mockClient.rbac());
```
Comment (tombentley) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperatorMockTest.java]: I don't think this is asserting anything useful. If you want to assert the existence of the SA you need to do the full query.
